File: go-backend-interview-prep/09_performance_and_profiling/02_benchmarks_basics/main.py
"""Benchmarks basics -- Python equivalent of the Go example.

Uses timeit (stdlib) to compare string building approaches.
"""

import timeit


def concat_strings(n: int) -> str:
    """Slow: += in a loop creates a new string each time."""
    s = ""
    for _ in range(n):
        s += "a"
    return s


def join_strings(n: int) -> str:
    """Fast: list + join allocates once at the end."""
    parts = []
    for _ in range(n):
        parts.append("a")
    return "".join(parts)


def main() -> None:
    n = 50_000

    print("=== String concat vs ''.join() ===")
    print(f"Building a {n}-char string...\n")

    # timeit: run each approach multiple times for stable results
    iters = 10

    concat_time = timeit.timeit(lambda: concat_strings(n), number=iters)
    join_time = timeit.timeit(lambda: join_strings(n), number=iters)

    print(f"  += concat:   {concat_time/iters*1000:.2f} ms/op  ({iters} iterations)")
    print(f"  ''.join():   {join_time/iters*1000:.2f} ms/op  ({iters} iterations)")
    print()
    print("Note: Python's timeit is the equivalent of Go's testing.B")
    print("  - timeit disables GC by default for stable timing")
    print("  - You set iteration count manually (Go auto-adjusts b.N)")
    print()
    print("For structured benchmarks, run:")
    print("  python ./09_performance_and_profiling/02_benchmarks_basics/bench_test.py")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/09_performance_and_profiling/02_benchmarks_basics/main.go
package main

import (
	"fmt"
	"strings"
	"time"
)

// concatStrings uses += in a loop (slow, many allocations).
func concatStrings(n int) string {
	s := ""
	for i := 0; i < n; i++ {
		s += "a"
	}
	return s
}

// buildStrings uses strings.Builder (fast, one allocation).
func buildStrings(n int) string {
	var b strings.Builder
	b.Grow(n)
	for i := 0; i < n; i++ {
		b.WriteByte('a')
	}
	return b.String()
}

func main() {
	n := 50_000

	fmt.Println("=== String concat vs strings.Builder ===")
	fmt.Printf("Building a %d-char string...\n\n", n)

	// Concat
	start := time.Now()
	s1 := concatStrings(n)
	concatTime := time.Since(start)

	// Builder
	start = time.Now()
	s2 := buildStrings(n)
	builderTime := time.Since(start)

	fmt.Printf("  += concat:       %s  (len=%d)\n", concatTime, len(s1))
	fmt.Printf("  strings.Builder: %s  (len=%d)\n", builderTime, len(s2))
	fmt.Println()
	fmt.Println("Note: This is a rough timing. For accurate results, use:")
	fmt.Println("  go test -bench=. -benchmem ./09_performance_and_profiling/02_benchmarks_basics/")
	fmt.Println()
	fmt.Println("Benchmark output format:")
	fmt.Println("  BenchmarkConcat-8   1000   150000 ns/op   1234567 B/op   49999 allocs/op")
	fmt.Println("                       ^N     ^time          ^bytes         ^heap allocs")
}
